src/plugins/irc-events/link.js

```diff
@@ -55,51 +58,108 @@ module.exports = function(client, chan, msg) {
 	});
 };
 
-function parse(msg, preview, res, client) {
-	switch (res.type) {
-	case "text/html": {
+function parseHtml(preview, res, client) {
+	return new Promise((resolve) => {
 		const $ = cheerio.load(res.data);
-		preview.type = "link";
-		preview.head =
-			$('meta[property="og:title"]').attr("content")
-			|| $("title").text()
-			|| "";
-		preview.body =
-			$('meta[property="og:description"]').attr("content")
-			|| $('meta[name="description"]').attr("content")
-			|| "";
-		preview.thumb =
-			$('meta[property="og:image"]').attr("content")
-			|| $('meta[name="twitter:image:src"]').attr("content")
-			|| $('link[rel="image_src"]').attr("href")
-			|| "";
-
-		if (preview.thumb.length) {
-			preview.thumb = url.resolve(preview.link, preview.thumb);
-		}
 
-		// Make sure thumbnail is a valid url
-		if (!/^https?:\/\//.test(preview.thumb)) {
-			preview.thumb = "";
-		}
+		return parseHtmlMedia($, preview, res, client)
+			.then((newRes) => resolve(newRes))
+			.catch(() => {
+				preview.type = "link";
+				preview.head =
+					$('meta[property="og:title"]').attr("content")
+					|| $("title").text()
+					|| "";
+				preview.body =
+					$('meta[property="og:description"]').attr("content")
+					|| $('meta[name="description"]').attr("content")
+					|| "";
+				preview.thumb =
+					$('meta[property="og:image"]').attr("content")
+					|| $('meta[name="twitter:image:src"]').attr("content")
+					|| $('link[rel="image_src"]').attr("href")
+					|| "";
+
+				if (preview.thumb.length) {
+					preview.thumb = url.resolve(preview.link, preview.thumb);
+				}
 
-		// Verify that thumbnail pic exists and is under allowed size
-		if (preview.thumb.length) {
-			fetch(escapeHeader(preview.thumb), {language: client.language}, (resThumb) => {
-				if (resThumb === null
-				|| !(/^image\/.+/.test(resThumb.type))
-				|| resThumb.size > (Helper.config.prefetchMaxImageSize * 1024)) {
+				// Make sure thumbnail is a valid url
+				if (!linkRegex.test(preview.thumb)) {
 					preview.thumb = "";
 				}
 
-				handlePreview(client, msg, preview, resThumb);
+				// Verify that thumbnail pic exists and is under allowed size
+				if (preview.thumb.length) {
+					fetch(escapeHeader(preview.thumb), client, (resThumb) => {
```

Comment:
`{language: client.language}` here?

Reply:
Just passing `client` should work as they variable has the same name, it gets destructed in the function anyway.

Reply:
Yeah but that introduces an implicit contract between the Client interface and the fetch function signature. This would not be the case if passing an object literal (this was initially pointed out by astorije in the PR where I added language support)

Reply:
Fair.